src/io/sb3/fromSb3.ts

```diff
@@ -85,93 +90,104 @@ function getBlockScript(blocks: { [key: string]: sb3.Block }) {
+                    return asNum;
+                  }
+                  return value;
                 };
+
+                switch (part) {
+                  case "%s":
+                  case "%n":
+                    return {
+                      type: "numberOrString",
+                      name: argNames.splice(0, 1)[0],
+                      defaultValue: optionalToNumber(argDefaults.splice(0, 1)[0])
+                    };
+                  case "%b":
+                    return {
+                      type: "boolean",
+                      name: argNames.splice(0, 1)[0],
+                      defaultValue: argDefaults.splice(0, 1)[0] === "true"
+                    };
+                  default:
+                    return {
+                      type: "label",
+                      name: part
+                    };
+                }
+              });
+
+              addInput("PROCCODE", { type: "string", value: mutation.proccode });
+              addInput("ARGUMENTS", { type: "customBlockArguments", value: args });
+              addInput("WARP", { type: "boolean", value: mutation.warp === "true" });
+            } else {
+              // In most cases, just copy the shadow block's fields and inputs
+              // into its parent
+              result = {
+                ...result,
+                ...translateInputs(blocks[value].inputs),
+                ...translateFields(blocks[value].fields, blocks[value].opcode)
+              };
+            }
+          } else {
+            let isBlocks;
+            if (BlockBase.isKnownBlock(block.opcode)) {
+              const { type } = BlockBase.getDefaultInput(block.opcode, inputName) || {};
+              if (type === "blocks") {
+                isBlocks = true;
               }
+            }
+            isBlocks = isBlocks || inputScript.length > 1;
+            if (isBlocks) {
+              addInput(inputName, { type: "blocks", value: inputScript });
             } else {
               addInput(inputName, { type: "block", value: inputScript[0] });
             }
-          } else {
-            addInput(inputName, { type: "blocks", value: inputScript });
           }
         } else if (value === null) {
           addInput(inputName, { type: "string", value: null });
         } else {
+          const BIS = sb3.BlockInputStatus;
           switch (value[0]) {
-            case 4:
-            case 5:
-            case 6:
-            case 7:
-              addInput(inputName, {
-                type: "number",
-                value: isNaN(parseFloat(value[1])) ? (value[1] as string) : parseFloat(value[1])
-              });
+            case BIS.MATH_NUM_PRIMITIVE:
+            case BIS.POSITIVE_NUM_PRIMITIVE:
+            case BIS.WHOLE_NUM_PRIMITIVE:
+            case BIS.INTEGER_NUM_PRIMITIVE: {
+              let storedValue: string | number = value[1];
+              if (!isNaN(parseFloat(storedValue as string))) {
```

Comment:
`parseFloat` [will parse things like `"1x"` as `1`](https://stackoverflow.com/questions/12227594/which-is-better-numberx-or-parsefloatx). Not sure if that's what Scratch does in this situation (or if that path will ever be hit), but it may be better to use a `Number` cast.